### PR TITLE
revised elasticsearch getting started guide

### PR DESCRIPTION
**Pull request by ackerman80**
In elasticsearch indices are a sensitive object to enforce access control.
This example better captures the security concerns for elasticsearch.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4364)

### CODE REVIEW

**Comment by ackerman80** [comment]
test-docs-please

**Comment by ackerman80** [comment]
test-me-please

File: Documentation/gettingstarted/elasticsearch.rst [hunk 1]
```diff
@@ -29,153 +34,156 @@ The file ``es-sw-app.yaml`` will deploy the Elasticsearch service which stores S
...
+For Elasticsearch clusters the **least privilege security** challenge is to give clients access only to particular indices, and to limit the operations each client is allowed to perform on each index. In this example, the ``outpost`` Elasticsearch clients only need access to upload troop logs; and the ``empire-hq`` client only needs search access to both the indices.  From the security perspective, the outposts are weak spots and susceptible to be captured by the rebels. Once compromised, the clients can be used to search and manipulate the critical data in Elasticsearch. We can simulate this attach using the commands below but first lets run the commands for legitimate behavior for all the client services.
```
Comment (raybejjani) [inline Documentation/gettingstarted/elasticsearch.rst]: `we can simulate this attach`,  should be `attack`

File: examples/kubernetes-es/es-sw-app.yaml [hunk 2]
```diff
@@ -38,7 +38,7 @@ spec:
       serviceAccount: elasticsearch
       initContainers:
       - name: init-sysctl
-        image: docker.io/library/busybox
+        image: busybox
```
Comment (raybejjani) [inline examples/kubernetes-es/es-sw-app.yaml]: I think some of the `docker.io` domains were added to image paths to fix some use-cases. Was there a reason to remove it?

File: examples/kubernetes-es/es-sw-app.yaml [hunk 3]
```diff
@@ -109,23 +109,33 @@ subjects:
...
   labels:
-    name: sidious
-    role: lord
+    app: outpost
 spec:
   containers:
-  - name: bookreader
-    image: docker.io/cilium/esclient:v1
+  - name: esclient
+    image: arvindsoni80/esindexclient:v1
```
Comment (raybejjani) [inline examples/kubernetes-es/es-sw-app.yaml]: I think we have random person-attached images elsewhere, but it might be a little sketchy looking to use this image as part of a cilium GSG. Could you upload this to the cilium dockerhub account? (this applies to the other uses of this image)

**Comment by ackerman80** [comment]
tested on K8s 1.8.0, 1.9.4 and 1.10.0. just tested for app creation in K8s 1.7.5

File: examples/kubernetes-es/es-sw-app.yaml [hunk 4]
```diff
@@ -109,23 +109,33 @@ subjects:
...
   labels:
-    name: sidious
-    role: lord
+    app: outpost
 spec:
   containers:
-  - name: bookreader
-    image: docker.io/cilium/esclient:v1
+  - name: esclient
+    image: cilium/esindexclient:v1
```
Comment (aanm) [inline examples/kubernetes-es/es-sw-app.yaml]: Please change this to `docker.io/cilium/esindexclient:v1`

Comment (aanm) [inline examples/kubernetes-es/es-sw-app.yaml]: Please upload the Dockefile used to build this image probably to a directory in `cilium/demo-image` and create an automated build out of there

**Review by aanm — DISMISSED** [review]

File: examples/kubernetes-es/es-sw-app.yaml (hunk 2 above)
Comment (ianvernon) [inline examples/kubernetes-es/es-sw-app.yaml]: Is this change necessary?

Comment (ackerman80) [reply examples/kubernetes-es/es-sw-app.yaml]: unintentional side effect. edited

Comment (ackerman80) [reply examples/kubernetes-es/es-sw-app.yaml]: not intentional. changed back

File: examples/kubernetes-es/es-sw-app.yaml (hunk 3 above)
Comment (ackerman80) [reply examples/kubernetes-es/es-sw-app.yaml]: addressed

File: Documentation/gettingstarted/elasticsearch.rst (hunk 1 above)
Comment (ackerman80) [reply Documentation/gettingstarted/elasticsearch.rst]: good eye :). addressed

File: Documentation/gettingstarted/elasticsearch.rst [hunk 5]
```diff
@@ -14,12 +14,17 @@ minutes.
 Step 2: Deploy the Demo Application
 ===================================
 
-Now that we have Cilium running, we can deploy our demo Elasticsearch application. The demo application aligns with our tradition of using Star Wars-themed examples. We go back to the time when Vader had recently been converted to the Dark side. Darth Sidious wanted to share some of the books that he had authored. The books were stored in an Elasticsearch database and exposed via a service of the same name. Both Sidious and Vader accessed these books using python-based clients running in Kubernetes pods.
+Following the Cilium tradition, we will use a Star Wars-inspired example. The Empire has a large scale Elasticsearch cluster which is used for storing variety of data including:
```
Comment (ianvernon) [inline Documentation/gettingstarted/elasticsearch.rst]: storing a variety of data

File: Documentation/gettingstarted/elasticsearch.rst [hunk 6]
```diff
@@ -14,12 +14,17 @@ minutes.
 Step 2: Deploy the Demo Application
 ===================================
 
-Now that we have Cilium running, we can deploy our demo Elasticsearch application. The demo application aligns with our tradition of using Star Wars-themed examples. We go back to the time when Vader had recently been converted to the Dark side. Darth Sidious wanted to share some of the books that he had authored. The books were stored in an Elasticsearch database and exposed via a service of the same name. Both Sidious and Vader accessed these books using python-based clients running in Kubernetes pods.
+Following the Cilium tradition, we will use a Star Wars-inspired example. The Empire has a large scale Elasticsearch cluster which is used for storing variety of data including: 
 
-.. image:: images/cilium_es_gsg_topology.png
-   :scale: 40 %
+* ``index: troop_logs``: Stormtroopers performance logs collected from every outpost and used to identify and eliminate weak performers!
```
Comment (ianvernon) [inline Documentation/gettingstarted/elasticsearch.rst]: from every outpost which are used to identify...

File: Documentation/gettingstarted/elasticsearch.rst [hunk 7]
```diff
@@ -14,12 +14,17 @@ minutes.
...
+* ``index: spaceship_diagnostics``: Spaceships diagnostics data collected from every spaceship and used for R&D and improvement of the spaceships.
```
Comment (ianvernon) [inline Documentation/gettingstarted/elasticsearch.rst]: diagnostics data collected from every spaceship which is used for R&D...

File: Documentation/gettingstarted/elasticsearch.rst [hunk 8]
```diff
@@ -14,12 +14,17 @@ minutes.
...
+
+Every outpost has an Elasticsearch client service to upload the Stormtroopers logs. And every spaceship has a service to upload diagnostics. Similarly, the Empire headquarters has a service to search and analyze the troop logs and spaceship diagnostics data. Before we look into the security concerns, lets first create this application scenario in minikube.
```
Comment (ianvernon) [inline Documentation/gettingstarted/elasticsearch.rst]: let's

File: Documentation/gettingstarted/elasticsearch.rst [hunk 9]
```diff
@@ -29,153 +34,154 @@ The file ``es-sw-app.yaml`` will deploy the Elasticsearch service which stores S
...
+For Elasticsearch clusters the **least privilege security** challenge is to give clients access only to particular indices, and to limit the operations each client is allowed to perform on each index. In this example, the ``outpost`` Elasticsearch clients only need access to upload troop logs; and the ``empire-hq`` client only needs search access to both the indices.  From the security perspective, the outposts are weak spots and susceptible to be captured by the rebels. Once compromised, the clients can be used to search and manipulate the critical data in Elasticsearch. We can simulate this attack but first let's run the commands for legitimate behavior for all the client services.
```
Comment (ianvernon) [inline Documentation/gettingstarted/elasticsearch.rst]: We can simulate this attack, but first let's run...

File: Documentation/gettingstarted/elasticsearch.rst [hunk 10]
```diff
@@ -29,153 +34,154 @@ The file ``es-sw-app.yaml`` will deploy the Elasticsearch service which stores S
...
+                 'stats': '[CRITICAL] [ENGINE BURN @SPEED 5000 km/s] [CHANCE 80%]'}}
+    Searching for Stormtroopers Performance Logs
+    Got 1 Hits:
+    {'_index': 'troop_logs', '_type': 'log', '_id': '1', '_score': 1.0, \
+     '_source': {'outpost': 'Endor', 'datetime': '33 ABY 4AM DST', 'title': 'Endor Corps 1: Morning Drill', \
+                 'notes': '5100 PRESENT; 15 ABSENT; 130 CODE-RED BELOW PAR PERFORMANCE'}}
 
 
-Sidious has access to perform ``GET`` and ``PUT`` as expected. But see what happens when Vader has both ``GET`` and ``PUT`` access. Vader can and did completely modify the books! (Note the change in book titles from 'Convert Jedi to Dark Side: 101' to "Why Convert a Jedi!").
+Now imagine an outpost captured by the rebels. In below commands the rebels first search all the indices and then manipulate the diagnostics data from a compromised outpost.
```
Comment (ianvernon) [inline Documentation/gettingstarted/elasticsearch.rst]: In the commands below, the rebels...

File: Documentation/gettingstarted/elasticsearch.rst [hunk 11]
```diff
@@ -29,153 +34,154 @@ The file ``es-sw-app.yaml`` will deploy the Elasticsearch service which stores S
...
+* ``outpost`` service only has upload access to ``index: troop_logs``
+* ``spaceship`` service only has upload access to ``index: spaceship_diagnostics``
+* ``empire-hq`` service only has search access for both the indices 
 
-In this example, the policy rules are defined for inbound traffic (i.e., "ingress") connections to the *elasticsearch* service. Note that endpoints selected as backend pods for the service are defined by the *selector* labels. *Selector* labels use the same concept as Kubernetes to define a service. In this example, label ``'k8s:component': elasticsearch`` defines the pods that are part of the *elasticsearch* service in Kubernetes.
+Fortunately, the Empire DevOps team is using Cilium for their Kubernetes cluster. Cilium provides L7 visibility and security policies to control Elasticsearch API access. Cilium follows the **white-list, least privilege model** for security. That is to say, a *CiliumNetworkPolicy* contains a list of rules that define **allowed requests** and any request that does math the rules is denied.
```
Comment (ianvernon) [inline Documentation/gettingstarted/elasticsearch.rst]: does not match the rules ?

File: Documentation/gettingstarted/elasticsearch.rst [hunk 12]
```diff
@@ -29,153 +34,154 @@ The file ``es-sw-app.yaml`` will deploy the Elasticsearch service which stores S
...
 
 .. parsed-literal::
 
     $ kubectl create -f \ |SCM_WEB|\/examples/kubernetes-es/es-sw-policy.yaml
-    ciliumnetworkpolicy "secure-empire-es" created
-
+    ciliumnetworkpolicy "secure-empire-elasticsearch" created
 
-Testing the security policy, Sidious still has both the ``GET`` and ``PUT`` access. But Vader now only has ``GET`` access and any attempts to ``PUT`` results in access denied.
+Testing the security policies, firstly, the search access is blocked for both outpost and spaceship. So from a compromised outpost, rebels will not be able to search and obtain knowledge about troops and spaceship diagnostics. Secondly, the outpost clients don't have access to create or update the ``index: spaceship_diagnostics``.
```
Comment (ianvernon) [inline Documentation/gettingstarted/elasticsearch.rst]: First sentence has awkward phrasing. Maybe:

"Let's test the security policies. Firstly, the search access..."

**Review by ianvernon — DISMISSED** [review]
Second pass - some grammatical / phrasing fixes are needed.

File: Documentation/gettingstarted/elasticsearch.rst (hunk 11 above)
Comment (ackerman80) [reply Documentation/gettingstarted/elasticsearch.rst]: @ianvernon what is not matching the rules?

Comment (ackerman80) [reply Documentation/gettingstarted/elasticsearch.rst]: nevermind got it

**Review by aanm — APPROVED** [review]

**Comment by ackerman80** [comment]
Do I need to do something more here or is this good for merging?

**Comment by ianvernon** [comment]
test-docs-please

**Review by ianvernon — APPROVED** [review]

**Comment by ianvernon** [comment]
Docs build passed, and all comments have been addressed; merging.
